packages/app/src/server/routes/api/cortexRuntimeConfig.ts

```diff
@@ -14,31 +14,90 @@
  * limitations under the License.
  */
 
-// import env from "server/env";
-import { log } from "@opstrace/utils/lib/log";
+import yaml from "js-yaml";
 import { NextFunction, Request, Response } from "express";
 
+import { log } from "@opstrace/utils/lib/log";
+
+import {
+  createOrUpdateConfigMapWithRetry,
+  ConfigMap
+} from "@opstrace/kubernetes";
+
+import { KubeConfig } from "@kubernetes/client-node";
+
 import { KUBECONFIG } from "../../server";
 
+function genCortexRuntimeConfigCM(kubeconfig: KubeConfig, yamldoc: string) {
+  // name and namespace and data key: are all convention-based, must be in sync
+  // with controller / installer.
+  return new ConfigMap(
+    {
+      apiVersion: "v1",
+      data: {
+        "runtime-config.yaml": yamldoc
+      },
+      kind: "ConfigMap",
+      metadata: {
+        name: "cortex-runtime-config",
+        namespace: "cortex",
+        annotations: {
+          // Custom convention, so that the Opstrace controller will not
+          // delete/overwrite this config map when it detects change.
+          opstrace: "no-update"
```

Comment:
instead of adding this here directly (avoiding the public APIs for immutability), it's better to create the configmap and then call `.setImmutable()` on the instance. The public methods for immutability should be `.setImmutable()` and `.isImmutable()`.

```
const cm = new ConfigMap({});
cm.setImmutable();
return cm;
```

Reply:
Sure, doing this :shrug: A little bit of context: I was questioning the 'public' aspect of `setImmutable()` :). I understand that it's a 'public method' on the 'ConfigMap' object.

Buuut: I was thinking that `opstrace: "no-update"` is not just a hidden unimportant implementation detail, but should maybe actually be the convention (public interface definition), and  `setImmutable()` would then in the context of lib/kubernetes make use of that convention :).

I think we should document the meaning of special key/value pairs in k8s metadata, and then our k8s objects would have a well-defined meaning independent of the tooling / library in use :). I imagine a _reference doc_ explaining in English what specific annotations mean, vs. having to read code about what they mean. Then this reference doc would describe a public interface.

I think this would be good for debugging, good for developer experience in general. Annotations returned by e.g. `kubectl` would raise the question: what do they mean? The answer would be in said document.

Now that I change this to `cm.setImmutable()` I am also adding a docstring to the method. From the non-Mat perspective  a `cm.setImmutable()` alone (w/o docstring) is not revealing what it does otherwise :).

Another thought, not important.. for methods like `setImmutable()` and let's maybe rethink the return value, maybe allowing for chaining.